### ISSUE TITLE
Promises

### ISSUE THREAD

**Issue by ronag** [nats-io/node-nats#24]
As far as I know all javascript platforms now support Promises natively. I would like to suggest that we upgrade the nats api to make use of that. Currently it doesn't even properly follow the node thunk pattern.

**Comment by derekcollison** [comment]
Apologies, what is the node thunk pattern? Promises were very early on when Ryan originally designed Node and then removed. Seems they are now back in, which is fine. What do you propose API wise? What part of the system would you want the promises to effect?

Thanks!

**Comment by ronag** [comment]
So in node, the pattern of callbacks is `function(error, result)` where the first argument is always the error and the second argument is the result. The way it is now makes the nats callbacks not compatible (without wrappers) with quite a lot of async libraries, such as highlandjs, rxjs etc...

I would suggest the following pattern:

````
Client.prototype.request = function(subject, opt_msg, opt_options, callback) {
  if (typeof opt_msg === 'function') {
    callback = opt_msg;
    opt_msg = EMPTY;
    opt_options = null;
  }
  if (typeof opt_options === 'function') {
    callback = opt_options;
    opt_options = null;
  }
  var inbox = createInbox();
  var s = this.subscribe(inbox, opt_options, function(msg, reply) {
    callback(null, {msg, reply});
  });
  this.publish(subject, opt_msg, inbox);
  return s;
};
```

**Comment by ronag** [comment]
Promise won't work with the current API. Will break to much since some methods return sids.

**Comment by derekcollison** [comment]
Not sure returning a sid gets rid of the possibility for promises being layered on top.

**Comment by ronag** [comment]
Care to elaborate?

**Comment by derekcollison** [comment]
Walking through your initial assumptions/suggestions.

request: returns promise if callback == null
**request is a stream, so not sure it's applicable.**

timeout: returns promise if callback == null
**Still possible**

subscribe: uses callback
**It has one now, but I assume you want args reordered?**

publish: remove unnecessary callback
**The callback is optional and allows code to execute after the PUB has been received by the server, which is very helpful**

flush: returns promise if callback == null
**Still possible IMO**

**Comment by ronag** [comment]
> request is a stream, so not sure it's applicable.

Yes, an Observable would be more appropriate.

> It has one now, but I assume you want args reordered?

Yes, and {msg, reply} should probably be combined into a single argument.

> The callback is optional and allows code to execute after the PUB has been received by the server, which is very helpful

This is a bit out of topic. But care to elaborate when that is helpful?

**Comment by derekcollison** [comment]
Why do you want msg and reply to a subscription's callback combined? Every message will have a subject, but reply is optional. 

Even though NATS is a fire and forget system, there are times when you want to make sure a message has been processed by the server within the application. The optional callback to publish() provides that ability.

**Comment by ronag** [comment]
> Why do you want msg and reply to a subscription's callback combined? Every message will have a subject, but reply is optional.

Because you cannot have multiple arguments in a thunk or promise.

**Comment by derekcollison** [comment]
I think subscriptions have a callback and are not converted to promises, IMO.

**Comment by ronag** [comment]
let me rephrase, a nodejs stream or Observable work on single item streams. But you are right, with a callback it doesn't really matter.

**Comment by haywirez** [comment]
Is there any way to handle queues using async functions? I have a bunch of asynchronous tasks that I'd like to have resolved before processing the next item in the queue.

**Comment by StarpTech** [comment]
Hi, I think you are looking for the modern gnats client https://github.com/nats-io/ts-nats

**Comment by haywirez** [comment]
@StarpTech thanks for the link, but I'm rather looking to figure out whether there's a way to use Promises/async functions in queue group subscribers, in order to await a task to finish before ingesting the next message. Something like:
```
nats.subscribe('foo', {'queue':'job.workers'}, async function() {
     // await async task to finish before taking next item in the queue...
});
```

**Comment by StarpTech** [comment]
@haywirez I think you are looking for an [async-generator`](https://zaiste.net/nodejs_10_asynchronous_iteration_async_generators/) implementation. That would be a great improvement but it's limited to Node 10.

```ts
for await (const msg of nats.subscribe('foo', {'queue':'job.workers'})) {
 // some code
}
```

**Comment by StarpTech** [comment]
@haywirez that's up to you. You can use https://github.com/delvedor/workq to buffer your work with a limited concurrency level.

**Comment by aricart** [comment]
The main issue here is that the inbound queue will be growing - depending on the rate messages are coming in, the memory will grow. Sounds to me that messages should be collected and then played back at a worker's pace - take a look at https://github.com/nats-io/nats-streaming-server
